src/cli/program.ts

```diff
@@ -1,130 +1,128 @@
-  );
-
-  if (unignoredTscErrors.length > 0) {
-    cliDependencies.log(
-      `${chalk.red(unignoredTscErrors.length)} errors were not ignored`,
-    );
+  private buildCompilationResultsInput(): ProcessCompilateResultsInput {
+    return {
+      cliDependencies: this.cliDependencies,
+      looselyTypeCheckedFilePathsArray: this.looselyTypeCheckedFilePathsArray,
+      looselyTypeCheckedFilePathMatcher: this.looselyTypeCheckedFilePathMatcher,
+      ignoredErrorCodes: this.ignoredErrorCodes,
+      tscErrors: this.errorStore.getAllErrors(),
+      ignoredTscErrors: this.errorStore.getIgnoredErrors(),
+      reportedTscErrors: this.errorStore.getReportedErrors(),
+      validTscErrors: this.errorStore.getValidErrors(),
+      tscErrorsThatCouldBeIgnored: this.errorStore.getCouldBeIgnoredErrors(),
+    };
   }
 
-  const updatedLooselyTypeCheckedFilePaths = new Set(
-    looselyTypeCheckedFilePathsArray,
-  );
-  const updatedIgnoredErrorCodes = new Set(ignoredErrorCodes);
-
-  const reportingResult = aggregateReportingResults([
-    reportTscErrorsThatCouldBeIgnored(
-      cliDependencies,
-      tscErrorsThatCouldBeIgnored,
-      updatedLooselyTypeCheckedFilePaths,
-    ),
-    reportLooselyTypeCheckedFilePathsWithoutErrors(
-      cliDependencies,
-      looselyTypeCheckedFilePathMatcher,
-      updatedLooselyTypeCheckedFilePaths,
-      tscErrors,
-    ),
-    reportValidTscErrors(cliDependencies, validTscErrors),
-    reportIgnoredErrorsThatDidNotOccur(
-      cliDependencies,
-      tscErrors,
-      updatedIgnoredErrorCodes,
-    ),
-  ]);
-
-  if (cliDependencies.cliOptions['auto-update']) {
-    updateLooselyTypeCheckedFilePaths(
-      cliDependencies,
-      new Set(looselyTypeCheckedFilePathsArray),
-      updatedLooselyTypeCheckedFilePaths,
-    );
-    updateIgnoredErrorCodes(
-      cliDependencies,
-      ignoredErrorCodes,
-      updatedIgnoredErrorCodes,
-    );
+  public processLine(line: string): void {
+    if (this.watchModeSpy.detectWatchModeCompilationStartOutput(line)) {
+      this.cliDependencies.log(
+        this.watchModeSpy.watchModeCompilationStartOuput,
+      );
+      return;
+    }
+
+    if (this.watchModeSpy.detectWatchModeCompilationFinishedOutput(line)) {
+      const compilationResultsInput = this.buildCompilationResultsInput();
+      processCompilationResults(compilationResultsInput);
+      this.cliDependencies.log(this.watchModeSpy.watchModeCompilationEndOutput);
+      return;
+    }
+
+    if (this.watchModeSpy.detectWatchModeFileChangeDetectedOutput(line)) {
+      this.cliDependencies.log(
+        this.watchModeSpy.watchModeFileChangeDetectedOutput,
+      );
+
+      this.lastTscError = undefined;
+      this.errorStore.reset();
+    }
+
+    const tscError = parseTscLine(line, this.lastTscError);
+
+    if (tscError !== this.lastTscError) {
+      const errorClassification = classifyTscError(
+        tscError as TscError,
```

Comment:
We could avoid a potentially-unsafe type assertion by adding a `&& tscError !== undefined` in the condition of the `if` above. I don't think it would hurt the `if` in any way, and it would make it possible to remove the 2 `as` type assertions here and 5 lines lower, which I'm always happy to do